java-checks/src/main/java/org/sonar/java/checks/BooleanInversionCheck.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ParenthesizedTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S1940",
+  name = "Boolean checks should not be inverted",
+  tags = {"pitfall"},
+  priority = Priority.MAJOR)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.UNDERSTANDABILITY)
+@SqaleConstantRemediation("2min")
+public class BooleanInversionCheck extends SubscriptionBaseVisitor {
+
+  private static Map<String, String> OPERATORS = ImmutableMap.<String, String>builder()
+    .put("==", "!=")
+    .put("!=", "==")
+    .put("<", ">=")
+    .put(">", "<=")
+    .put("<=", ">")
+    .put(">=", "<")
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.LOGICAL_COMPLEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ExpressionTree expression = removeParenthesis(((UnaryExpressionTree) tree).expression());
+    // The logic could be simplified if kind would be exposed in the ST
+    if (expression.is(
+        Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO,
+        Tree.Kind.LESS_THAN, Tree.Kind.GREATER_THAN,
+        Tree.Kind.LESS_THAN_OR_EQUAL_TO, Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
+      context.addIssue(tree, this, "Use the opposite operator (\"" + OPERATORS.get(((BinaryExpressionTree) expression).operatorToken().text()) + "\") instead.");
+    }
+  }
+
+  private static ExpressionTree removeParenthesis(ExpressionTree tree) {
```

Comment:
This one should really make its way (in a dedicated commit) into a utility class for now. It has already been implemented more than necessary.